ion_schema_2_0/constraints/regex.isl

```diff
@@ -0,0 +1,1091 @@
+    "hello	world",
+  ],
+  should_reject_as_invalid: [
+    null,
+    null.null,
+    null.string,
+    null.symbol,
+    null.bool,
+    'hello world',
+    "hello world",
+  ]
+}
+
+type::{
+  name: regex_grouping,
+  regex: "123(ab|cd|ef)456",
+}
+$test::{
+  type: regex_grouping,
+  should_accept_as_valid: [
+    "123ab456",
+    "123cd456",
+    "123ef456",
+  ],
+  should_reject_as_invalid: [
+    "123a456",
+    "123ac456",
+    "123ace456",
+    "123bdf456",
+  ]
+}
+
+type::{
+  name: regex_ion_escape_characters,
+  regex: "\0\a\b\t\n\f\r\v",
+}
+$test::{
+  type: regex_ion_escape_characters,
+  should_accept_as_valid: [
+    '\0\a\b\t\n\f\r\v',
+    "\0\a\b\t\n\f\r\v",
+  ],
+  should_reject_as_invalid: [
+    null,
+    null.null,
+    null.string,
+    null.symbol,
+    null.bool,
+    'hello world',
+    "hello world",
+  ]
+}
+
+type::{
+  name: regex_multiline,
+  regex: m::"^hello world$",
+}
+$test::{
+  type: regex_multiline,
+  should_accept_as_valid: [
+    'hello world',
+    'hello world\n\n\n',
+    '\n\n\nhello world',
+    '\n\n\nhello world\n\n\n',
+    '\r\r\rhello world\r\r\r',
+
+    "hello world",
+    "hello world\n\n\n",
+    "\n\n\nhello world",
+    "\n\n\nhello world\n\n\n",
+    "\r\r\rhello world\r\r\r",
+  ],
+  should_reject_as_invalid: [
+    null,
+    null.null,
+    null.string,
+    null.symbol,
+    null.bool,
+    " hello world",
+    "hello world ",
+    "\n hello world",
+    "hello world \n",
+    "hello\n world",
+    "hello \nworld",
+    "hello\r world",
+    "hello \rworld",
+  ]
+}
+
+type::{
+  name: regex_predefined_character_class_digit,
+  regex: "\\d",
+}
+$test::{
+  type: regex_predefined_character_class_digit,
+  should_accept_as_valid: [
+    "0",
+    "1",
+    "2",
+    "3",
+    "4",
+    "5",
+    "6",
+    "7",
+    "8",
+    "9",
+  ],
+  should_reject_as_invalid: [
+    "/",
+    ":",
+  ]
+}
+
+type::{
+  name: regex_predefined_character_class_non_digit,
+  regex: "\\D",
+}
+$test::{
+  type: regex_predefined_character_class_non_digit,
+  should_accept_as_valid: [
+    "/",
+    ":",
+  ],
+  should_reject_as_invalid: [
+    "0",
+    "1",
+    "2",
+    "3",
+    "4",
+    "5",
+    "6",
+    "7",
+    "8",
+    "9",
+  ]
+}
+
+type::{
+  name: regex_predefined_character_class_non_whitespace,
+  regex: "\\S",
+}
+$test::{
+  type: regex_predefined_character_class_non_whitespace,
+  should_accept_as_valid: [
+    "\0",
+    "\b",
+    "\v",
+    "a",
+    "9",
+  ],
+  should_reject_as_invalid: [
+    " ",
+    "\f",
+    "\n",
+    "\r",
+    "\t",
+  ]
+}
+
+type::{
+  name: regex_predefined_character_class_non_word,
+  regex: "\\W",
+}
+$test::{
+  type: regex_predefined_character_class_non_word,
+  should_accept_as_valid: [
+    "/",
+    ":",
+    "@",
+    "[",
+    "`",
+    "{",
+  ],
+  should_reject_as_invalid: [
+    "a",
+    "A",
+    "z",
+    "Z",
+    "0",
+    "9",
+  ]
+}
+
+type::{
+  name: regex_predefined_character_class_whitespace,
+  regex: "\\s",
```

Comment:
What if `regex` could also accept a symbol instead of [just a string](https://amzn.github.io/ion-schema/docs/isl-2-0/spec#regex)? Wouldn't that make it easier to specify regexes, which often include `\`?

Reply:
No it would not make much (any?) difference, because the `\` still needs to be escaped inside of a symbol so that you can have a symbol contain the `'` character.